### PR TITLE
AMDGPU: Handle multiple AGPR MFMA rewrites

### PR DESCRIPTION
**Pull request by arsenm**
I have this firing on one of the real examples, need to
produce the tests and check a few edge cases

### CODE REVIEW

**Comment by arsenm** [comment]
* **#150566** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149027** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/149027?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149026** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/149026?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#147975** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/147975?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/147975?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by srpande — COMMENTED** [review]
What if src2 is 0?

```
DefMI %631:av_512_align2 = COPY %632:vreg_512_align2

CopySrcMI %632:vreg_512_align2 = contract V_MFMA_F32_32X32X2F32_vgprcd_e64 %628.sub0:av_128_align2, %234.sub0:vreg_128_align2, **0**, 0, 0, 0, implicit $mode, implicit $exec
```
This will crash.

**Comment by arsenm** [comment]
> What if src2 is 0?
> 
> ```
> DefMI %631:av_512_align2 = COPY %632:vreg_512_align2
> 
> CopySrcMI %632:vreg_512_align2 = contract V_MFMA_F32_32X32X2F32_vgprcd_e64 %628.sub0:av_128_align2, %234.sub0:vreg_128_align2, **0**, 0, 0, 0, implicit $mode, implicit $exec
> ```
> 
> This will crash.

It's broken, but it was already broken before this patch

**Comment by arsenm** [comment]
> > What if src2 is 0?
> > ```
> > DefMI %631:av_512_align2 = COPY %632:vreg_512_align2
> > 
> > CopySrcMI %632:vreg_512_align2 = contract V_MFMA_F32_32X32X2F32_vgprcd_e64 %628.sub0:av_128_align2, %234.sub0:vreg_128_align2, **0**, 0, 0, 0, implicit $mode, implicit $exec
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This will crash.
> 
> It's broken, but it was already broken before this patch

https://github.com/llvm/llvm-project/pull/150552

File: llvm/lib/Target/AMDGPU/AMDGPURewriteAGPRCopyMFMA.cpp [hunk 1]
```diff
@@ -208,8 +218,19 @@ bool AMDGPURewriteAGPRCopyMFMAImpl::run(MachineFunction &MF) const {
 
       CopySrcMI->setDesc(TII.get(AGPROp));
 
-      // TODO: Is replacing too aggressive, fixup these instructions only?
-      MRI.replaceRegWith(CopySrcReg, VReg);
+      // Perform replacement of the register, rewriting the rewritable uses.
+      for (MachineInstr &UseMI :
```
Comment (perlfu) [inline llvm/lib/Target/AMDGPU/AMDGPURewriteAGPRCopyMFMA.cpp:222]: Is `UseMI` here misleading as `reg_instructions` will also include defs?

Comment (lucas-rami) [reply llvm/lib/Target/AMDGPU/AMDGPURewriteAGPRCopyMFMA.cpp:240]: I think it's a bit confusing too. The comment above also only mentions uses as being rewritten whereas defs will also have their register substituted.

**Review by lucas-rami — APPROVED** [review]
LGTM modulo the nit on `UseMI`.
